ops/model.py

```diff
@@ -1326,6 +1328,96 @@ def list_files(self, path: str, *, pattern: str = None,
         """
         return self._pebble.list_files(path, pattern=pattern, itself=itself)
 
+    def push_path(self, source_path: typing.Union[typing.Union[str, Path], typing.List[typing.Union[str, Path]]], dest_dir: typing.Union[str, Path]):
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        def local_list(source_path: str):
+            paths = [source_path]
+            if os.path.isdir(source_path):
+                paths = os.listdir(source_path)
+            files = []
+            for name in paths:
+                fpath = os.path.join(source_path, name)
+                info = os.lstat(fpath)
+                if stat.S_ISDIR(info.st_mode):
+                    ftype = pebble.FileType.DIRECTORY
+                elif stat.S_ISREG(info.st_mode):
+                    ftype = pebble.FileType.FILE
+                elif stat.S_ISLINK(info.st_mode):
+                    ftype = pebble.FileType.SYMLINK
+                else:
+                    continue
+                files.append(pebble.FileInfo(fpath, os.path.basename(fpath), ftype, info.st_size, stat.S_IMODE(info.st_mode), info.st_mtime, info.st_uid, pwd.getpwuid(info.st_uid).pw_name, info.st_gid, grp.getgrgid(info.st_gid).gr_name))
```

Comment:
Would this be cleaner if implemented with `map` and `os.walk`?

Reply:
I think that it might be?

If so, it might be worth it -- shoving some of the branching logic here off to a library would be appropriate.

Reply:
The code is designed to re-use the recursion between both push and pull operations as well as keep push/pull semantics as consistent as possible.  We can't do os.walk on the remote container - we have to recurse ourselves.  Also, this function isn't actually doing the recursion.  It just collects pertinent info about entries in a single directory.